app/decorators/controllers/solidus_gdpr/spree/api/users_controller_decorator.rb

```diff
@@ -0,0 +1,43 @@
+# frozen_string_literal: true
+
+module SolidusGdpr
+  module Spree
+    module Api
+      module UsersControllerDecorator
+        def emails
+          users_scope = model_class.accessible_by(current_ability, :read)
+          orders_scope = ::Spree::Order.all
+
+          users = users_scope.ransack(params[:q]).result.select(:email)
+          orders = orders_scope.ransack(params[:q]).result(distinct: true).select(:email)
+
+          results = ActiveRecord::Base.connection.execute <<~SQL
+            SELECT emails.* FROM (#{users.to_sql} UNION #{orders.to_sql}) emails
+          SQL
+
+          scope = results.map do |result|
+            {
+              id: result['email'],
+              email: result['email'],
+              orders_count: user_orders_count(result['email'])
+            }
+          end
+
+          @users = Kaminari.paginate_array(scope)
```

Comment:
We should paginate before loading the records. At this point, all of the emails will already have been loaded into memory. [This article](https://kirshatrov.com/2015/11/08/kaminari-custom-query/) provides some useful information on how to paginate raw SQL queries with Kaminari.